src/aws-codeartifact/repository.ts

```diff
@@ -0,0 +1,357 @@
+
+  /**
+   * TagManager to set, remove and format tags
+   */
+  readonly cdkTagManager: TagManager;
+  /**
+   * The ARN (Amazon Resource Name) of this CodeArtifact repository.
+   */
+  readonly repositoryArn: string;
+  /**
+   * The name of this CodeArtifact repository.
+   */
+  readonly repositoryName: string;
+  /**
+   * The domain that contains this repository.
+   */
+  readonly domain: IDomain;
+  /**
+   * Optional policy document that represents the resource policy of this repository
+   *
+   * If specified, addToResourcePolicy can be used to edit this policy.
+   * Otherwise this method will no-op.
+   */
+  protected policy?: PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: RepositoryProps) {
+    super(scope, id);
+
+    this.cdkTagManager = new TagManager(TagType.KEY_VALUE, 'AWS::CodeArtifact::Repository');
+
+    this.cfnResourceProps = {
+      domainName: props.domain.domainName,
+      repositoryName: props.repositoryName,
+      description: props.description,
+      domainOwner: props.domain.domainOwner,
+      externalConnections: props.externalConnection !== undefined ? [props.externalConnection] : undefined, // only 1 allowed
+      permissionsPolicyDocument: Lazy.any({ produce: () => this.policy?.toJSON() }),
+      upstreams: props.upstreams?.map(repo => repo.repositoryName),
+    };
+
+    this.cfnResource = this.createCfnResource();
+
+    this.repositoryArn = this.cfnResource.attrArn;
+    this.repositoryName = this.cfnResource.attrName;
+    this.domain = Domain.fromDomainArn(
+      this,
+      'domain',
+      Stack.of(this).formatArn({
+        resource: 'domain',
+        service: 'codeartifact',
+        account: this.cfnResource.attrDomainOwner,
+        resourceName: this.cfnResource.attrDomainName,
+      }),
+    );
+  }
+
+  protected createCfnResource(): CfnRepository {
+    return new CfnRepository(this, 'Resource', this.cfnResourceProps);
+  }
+
+  /**
+   * Adds a statement to the KMS key resource policy.
+   * @param statement The policy statement to add
+   */
+  public addToResourcePolicy(statement: PolicyStatement): AddToResourcePolicyResult {
+    if (!this.policy) {
+      this.policy = new PolicyDocument();
+    }
+
+    this.policy.addStatements(statement);
+    return { statementAdded: true, policyDependable: this.policy };
+  }
+
+  private isCrossEnvironmentGrantee(grantee: IGrantable): boolean {
+    if (!principalIsOwnedResource(grantee.grantPrincipal)) {
+      return false;
+    }
+    const thisStack = Stack.of(this);
+    const identityStack = Stack.of(grantee.grantPrincipal);
+    return thisStack.region !== identityStack.region || thisStack.account !== identityStack.account;
+  }
+
+  /**
+   * Grants permissions to the specified grantee on this CodeArtifact repository.
+   *
+   * @param grantee The principal to grant permissions to
+   * @param actions The actions to grant
+   */
+  public grant(grantee: IGrantable, ...actions: string[]): Grant {
```

Comment:
It would be good to define  the `grant` , `grantRead` and `grantReadAndPublish` methods in the `RepositoryBase` class to call the methods from imported resources.

Reply:
The `addToResourcePolicy` and `policy` variable could also be moved to the `RepositoryBase`. If we never call from imported resources, we don't need to define them in `IRepository`.

(We can also define it in `IRepository` and create a method that just returns the following in the import resource, but which is better would be a preference.)

```ts
  public grant(grantee: IGrantable, ...actions: string[]): Grant {
    return { statementAdded: false };
  }
```

Reply:
I have implemented it like this. Take a look :)

Reply:
Will do the same with `IDomain`

Reply:
Thanks for the changes, but some fixes appear to have been reverted to their old state. Could you please check?

- JSDoc for `policy?: PolicyDocument;` (line 184-189)
  - Also, there is the word `this key`(copy from the KMS module...?)
- `i.e. external key` (line 205)
- `allowNoOp` in `addToResourcePolicy` (line 208)
- `this.policy = new PolicyDocument();` (line 386)

Reply:
Yeah, there was still some copy-:spaghetti: left. It should all be fixed now. I hope.